File: SingletonStorage/LLMstore.py

import base64
from datetime import datetime
import io
from PIL import Image
from typing import List
from zoneinfo import ZoneInfo
from .BasicModel import BasicStore, Controller4Basic, Model4Basic

def get_current_datetime_with_utc():
    return datetime.now().replace(tzinfo=ZoneInfo("UTC"))

class Controller4LLM:
    class AbstractObjController(Controller4Basic.AbstractObjController):
        pass

    class CommonDataController(AbstractObjController):
        pass

    class AuthorController(AbstractObjController):
        pass

    class AbstractContentController(AbstractObjController):
        def __init__(self, store, model):
            self.model: Model4LLM.AbstractContent = model
            self._store:LLMstore = store

        def delete(self):
            self.get_data().get_controller().delete()
            # self.storage().delete_obj(self.model)
            super().delete()

        def get_author(self):
            author:Model4LLM.Author = self.storage().find(self.model.author_id)
            return author

        def get_group(self):
            res:Model4LLM.ContentGroup = self.storage().find(self.model.group_id)
            return res
        
        def get_data(self):
            res:Model4LLM.CommonData = self.storage().find(self.model.data_id())
            return res

        def get_data_raw(self):
            return self.get_data().raw    

        def update_data_raw(self, msg: str):
            self.get_data().get_controller().update(raw = msg)
            return self

        def append_data_raw(self, msg: str):
            data = self.get_data()
            return self.update_data_raw(data.raw + msg)
        
    class AbstractGroupController(Controller4Basic.AbstractGroupController):
        # def __init__(self, store, model):
        #     self.model:Model4LLM.AbstractGroup = model
        #     self._store:LLMstore = store

        # def yield_children_recursive(self, depth: int = 0):
        #     for child_id in self.model.children_id:
        #         if not self.storage().exists(child_id):
        #             continue
        #         content:Model4LLM.AbstractObj = self.storage().find(child_id)
        #         yield content, depth
        #         if child_id.startswith('ContentGroup'):
        #             group:Controller4LLM.AbstractGroupController = content._controller
        #             for cc, d in group.yield_children_recursive(depth + 1):
        #                 yield cc, d

        # def delete_recursive(self):
        #     for c, d in self.yield_children_recursive():
        #         c.get_controller().delete()
        #     self.delete()

        # def get_children(self):
        #     # self.load()
        #     assert  self.model is not None, 'controller has null model!'
        #     results:List[Model4LLM.AbstractObj] = []
        #     for child_id in self.model.children_id:
        #         results.append(self.storage().find(child_id))
        #     return results

        # def get_child(self, child_id: str):
        #     res:Model4LLM.AbstractContent = self.storage().find(child_id)
        #     return res

        def prints(self):
            res = '########################################################\n'
            for content, depth in self.yield_children_recursive():
                res += f"{'    ' * depth}{content.get_id()}\n"
            res += '########################################################\n'
            print(res)
            return res

    class ContentGroupController(AbstractGroupController):
        def __init__(self, store, model):
            self.model:Model4LLM.ContentGroup = model
            self._store:LLMstore = store
        
        def storage(self):return self._store

        def add_new_child_group(self,metadata={},rank=[0]):
            parent,child = self.storage().add_new_group_to_group(group=self.model,metadata=metadata,rank=rank)
            return child

        def add_new_text_content(self, author_id:str, text:str):
            parent,child = self.storage().add_new_text_to_group(group=self.model,author_id=author_id,
                                                    text=text)                             
            return child
        
        def add_new_embeding_content(self, author_id:str, content_id:str, vec:list[float]):
            parent,child = self.storage().add_new_embedding_to_group(group=self.model,author_id=author_id,
                                                        content_id=content_id, vec=vec)                                   
            return child
        
        def add_new_image_content(self,author_id:str, filepath:str):
            parent,child = self.storage().add_new_image_to_group(group=self.model,author_id=author_id,
                                                    filepath=filepath)                              
            return child
            

        # def delete_child(self, child_id:str):
        #     remaining_ids = [cid for cid in self.model.children_id if cid != child_id]
        #     for content in self.get_children():
        #         if content.get_controller().model.get_id() == child_id:
        #             if child_id.startswith('ContentGroup'):
        #                 group:Controller4LLM.ContentGroupController = content.get_controller()
        #                 group.delete_recursive()
        #             else:
        #                 content.get_controller().delete()
        #             break
        #     self.update(children_id = remaining_ids)
        #     return self

        # def get_children_recursive(self):
        #     results:list[Model4LLM.AbstractContent] = []
        #     for c, d in self.yield_children_recursive():
        #         results.append(c)
        #     return results

    class TextContentController(AbstractContentController):
        pass

    class EmbeddingContentController(AbstractContentController):
        def __init__(self, store, model):
            self.model: Model4LLM.EmbeddingContent = model
            self._store:LLMstore = store
            
        def get_data_raw(self):
            return list(map(float,super().get_data_raw()[1:-1].split(',')))
        
        def get_data_rLOD0(self):
            return self.get_data_raw()[::10**(0+1)]
        
        def get_data_rLOD1(self):
            return self.get_data_raw()[::10**(1+1)]
        
        def get_data_rLOD2(self):
            return self.get_data_raw()[::10**(2+1)]
        
        def get_target(self):
            assert  self.model is not None, 'controller has null model!'
            target_id = self.model.target_id
            res:Model4LLM.AbstractContent = self.storage().find(target_id)
            return res
        
        def update_data_raw(self, embedding: list[float]):
            super().update_data_raw(str(embedding))
            return self

    class FileLinkContentController(AbstractContentController):
        pass

    class BinaryFileContentController(AbstractContentController):
        def __init__(self, store, model):
            self.model: Model4LLM.BinaryFileContent = model
            self._store:LLMstore = store
            
        def read_bytes(self, filepath):
            with open(filepath, "rb") as f:
                return f.read()
            
        def b64decode(self, file_base64):
            return base64.b64decode(file_base64)
            
        def get_data_rLOD0(self):
            raise ValueError('binary file has no LOD concept')
        
        def get_data_rLOD1(self):
            raise ValueError('binary file has no LOD concept')
        
        def get_data_rLOD2(self):
            raise ValueError('binary file has no LOD concept')
        
    class ImageContentController(BinaryFileContentController):
        def __init__(self, store, model):
            self.model: Model4LLM.ImageContent = model    
            self._store:LLMstore = store

        def decode_image(self, encoded_string):
            return Image.open(io.BytesIO(self.b64decode(encoded_string)))
        
        def get_image(self):
            encoded_image = self.get_data_raw()
            if encoded_image:
                image = self.decode_image(encoded_image)
                return image
            return None
                    
        def get_image_format(self):
            image = self.get_image()
            return image.format if image else None
        
        def get_data_rLOD(self,lod=0):
            image = self.get_image()
            ratio = 10**(lod+1)
            if image.size[0]//ratio==0 or image.size[1]//ratio ==0:
                raise ValueError(f'img size({image.size}) of LOD{lod} is smaller than 0')
            return image.resize((image.size[0]//ratio,image.size[1]//ratio)) if image else None

        def get_data_rLOD0(self):
            return self.get_data_rLOD(lod=0)
        
        def get_data_rLOD1(self):
            return self.get_data_rLOD(lod=1)
        
        def get_data_rLOD2(self):
            return self.get_data_rLOD(lod=2)
    
class Model4LLM:
    class AbstractObj(Model4Basic.AbstractObj):
        def _get_controller_class(self,modelclass=Controller4LLM):
            class_type = self.__class__.__name__+'Controller'
            res = {c.__name__:c for c in [i for k,i in modelclass.__dict__.items() if '_' not in k]}
            res = res.get(class_type, None)
            if res is None: raise ValueError(f'No such class of {class_type}')
            return res

    class AbstractGroup(Model4Basic.AbstractGroup):
        _controller: Controller4LLM.AbstractGroupController = None
        def get_controller(self):return self._controller

        def _get_controller_class(self,modelclass=Controller4LLM):
            class_type = self.__class__.__name__+'Controller'
            res = {c.__name__:c for c in [i for k,i in modelclass.__dict__.items() if '_' not in k]}
            res = res.get(class_type, None)
            if res is None: raise ValueError(f'No such class of {class_type}')
            return res
        
    class CommonData(AbstractObj):
        raw: str = ''
        rLOD0: str = ''
        rLOD1: str = ''
        rLOD2: str = ''
        _controller: Controller4LLM.CommonDataController = None
        def get_controller(self):return self._controller

    class Author(AbstractObj):
        name: str = ''
        role: str = ''
        _controller: Controller4LLM.AuthorController = None
        def get_controller(self):return self._controller

    class AbstractContent(AbstractObj):
        author_id: str=''
        group_id: str=''
        _controller: Controller4LLM.AbstractContentController = None
        def get_controller(self):return self._controller
        def data_id(self):return f"CommonData:{self.get_id()}"
        
    class ContentGroup(AbstractGroup):
        _controller: Controller4LLM.ContentGroupController = None
        def get_controller(self):return self._controller

    class TextContent(AbstractContent):
        _controller: Controller4LLM.TextContentController = None
        def get_controller(self):return self._controller

    class EmbeddingContent(AbstractContent):
        target_id: str
        _controller: Controller4LLM.EmbeddingContentController = None
        def get_controller(self):return self._controller
        
    class FileLinkContent(AbstractContent):
        _controller: Controller4LLM.FileLinkContentController = None
        def get_controller(self):return self._controller

    class BinaryFileContent(AbstractContent):
        _controller: Controller4LLM.BinaryFileContentController = None
        def get_controller(self):return self._controller

    class ImageContent(BinaryFileContent):
        _controller: Controller4LLM.ImageContentController = None
        def get_controller(self):return self._controller


class LLMstore(BasicStore):    
    
    def __init__(self, version_controll=False) -> None:
        super().__init__(version_controll)
        self.python_backend()

    def _get_class(self, id: str, modelclass=Model4LLM):
        return super()._get_class(id, modelclass)
    
    def add_new_author(self,name, role, rank:list=[0], metadata={}) -> Model4LLM.Author:
        return self.add_new_obj(Model4LLM.Author(name=name, role=role, rank=rank, metadata=metadata))
    
    def add_new_root_group(self,metadata={},rank=[0]) -> Model4LLM.ContentGroup:
        return self.add_new_obj( Model4LLM.ContentGroup(rank=rank, metadata=metadata) )
    
    def _add_new_content_to_group(self,group:Model4LLM.ContentGroup,content:Model4LLM.AbstractContent,raw:str=None):
        if group._id is None:raise ValueError('the group is no exits!')
        content = self.add_new_obj(content)
        group.get_controller().update(children_id=group.children_id+[content.get_id()])
        if raw is not None and 'ContentGroup' not in content.get_id():
            self.add_new_obj(Model4LLM.CommonData(raw=raw),id=content.data_id())
        content.init_controller(self)
        return group,content    

    def read_image(self, filepath):
        with open(filepath, "rb") as f:
            return f.read()
        
    def b64encode(self, file_bytes):
        return base64.b64encode(file_bytes)
        
    def encode_image(self, image_bytes):
        return self.b64encode(image_bytes)
    
    def add_new_group_to_group(self,group:Model4LLM.ContentGroup,metadata={},rank=[0]):
        parent,child = self._add_new_content_to_group(group, Model4LLM.ContentGroup(rank=rank, metadata=metadata, parent_id=group.get_id()))
        return parent,child

    def add_new_text_to_group(self,group:Model4LLM.ContentGroup,author_id:str,text:str):
        parent,child = self._add_new_content_to_group(group,
                                                      Model4LLM.TextContent(author_id=author_id, group_id=group.get_id()),
                                                      raw=text)
        return parent,child
    
    def add_new_embedding_to_group(self,group:Model4LLM.ContentGroup, author_id:str, content_id:str, vec:list[float]):
        parent,child = self._add_new_content_to_group(group,
                                                      Model4LLM.EmbeddingContent(author_id=author_id, 
                                                                       group_id=group.get_id(),target_id=content_id),
                                                      raw=str(vec))
        return parent,child
    
    def add_new_image_to_group(self,group:Model4LLM.ContentGroup,author_id:str, filepath:str):
        raw_bytes = self.read_image(filepath)
        raw_base64 = self.encode_image(raw_bytes)
        parent,child = self._add_new_content_to_group(group,
                                                      Model4LLM.ImageContent(author_id=author_id,group_id=group.get_id()),
                                                      raw=raw_base64)
        return parent,child
    
    
    def find_group(self,id:str) -> Model4LLM.ContentGroup:
        assert 'Group:' in id, 'this id is not a id for goup'
        return self.find(id)
    
    def find_all_authors(self)->list[Model4LLM.Author]:
        return self.find_all('Author:*')

File: SingletonStorage/LLMstore.js

class Controller4LLM {
    static AbstractObjController = class extends Controller4Basic.AbstractObjController {
        // Inherits functionality from Controller4Basic.AbstractObjController
    };

    static CommonDataController = class extends Controller4LLM.AbstractObjController {
        constructor(store, model) {
            super(store, model);
            this.model = model;  // Model4LLM.CommonData
            this._store = store;  // LLMstore
        }
    };

    static AuthorController = class extends Controller4LLM.AbstractObjController {
        constructor(store, model) {
            super(store, model);
            this.model = model;  // Model4LLM.Author
            this._store = store;  // LLMstore
        }
    };

    static AbstractContentController = class extends Controller4LLM.AbstractObjController {
        constructor(store, model) {
            super(store, model);
            this.model = model;  // Model4LLM.AbstractContent
            this._store = store;  // LLMstore
        }

        delete() {
            this.get_data().get_controller().delete();
            this._store.delete(this.model.get_id());
            this.model._controller = null;
        }

        get_author() {
            return this._store.find(this.model.author_id);  // Model4LLM.Author
        }

        get_group() {
            return this._store.find(this.model.group_id);  // Model4LLM.ContentGroup
        }

        get_data() {
            return this._store.find(this.model.data_id());  // Model4LLM.CommonData
        }

        get_data_raw() {
            return this.get_data().raw;
        }

        update_data_raw(msg) {
            this.get_data().get_controller().update({ raw: msg });
            return this;
        }

        append_data_raw(msg) {
            const data = this.get_data();
            data.get_controller().update({ raw: data.raw + msg });
            return this;
        }
    };

    static AbstractGroupController = class extends Controller4LLM.AbstractObjController {
        constructor(store, model) {
            super(store, model);
            this.model = model;  // Model4LLM.AbstractGroup
            this._store = store;  // LLMstore
        }

        *yield_children_content_recursive(depth = 0) {
            for (const child_id of this.model.children_id) {
                if (!this._store.exists(child_id)) continue;
                const content = this._store.find(child_id);  // Model4LLM.AbstractObj
                yield [content, depth];
                if (child_id.startsWith('ContentGroup')) {
                    const group = content.get_controller();  // Controller4LLM.AbstractGroupController
                    for (const [cc, d] of group.yield_children_content_recursive(depth + 1)) {
                        yield [cc, d];
                    }
                }
            }
        }

        delete_recursive_from_key_value_storage() {
            for (const [content] of this.yield_children_content_recursive()) {
                content.get_controller().delete();
            }
            this.delete();
        }

        get_children_content() {
            return this.model.children_id.map(child_id => this._store.find(child_id));  // Array of Model4LLM.AbstractObj
        }

        get_child_content(child_id) {
            return this._store.find(child_id);  // Model4LLM.AbstractContent
        }

        prints() {
            let result = '########################################################\n';
            for (const [content, depth] of this.yield_children_content_recursive()) {
                result += `${'    '.repeat(depth)}${content.get_id()}\n`;
            }
            result += '########################################################\n';
            console.log(result);
            return result;
        }
    };

    static ContentGroupController = class extends Controller4LLM.AbstractGroupController {
        constructor(store, model) {
            super(store, model);
            this.model = model;  // Model4LLM.ContentGroup
            this._store = store;  // LLMstore
        }

        add_new_child_group(metadata = {}, rank = [0]) {
            const [, child] = this._store.add_new_group_to_group(this.model, metadata, rank);
            return child;
        }

        add_new_text_content(author_id, text) {
            const [, child] = this._store.add_new_text_to_group(this.model, author_id, text);
            return child;
        }

        add_new_embedding_content(author_id, content_id, vec) {
            const [, child] = this._store.add_new_embedding_to_group(this.model, author_id, content_id, vec);
            return child;
        }

        add_new_image_content(author_id, filepath) {
            const [, child] = this._store.add_new_image_to_group(this.model, author_id, filepath);
            return child;
        }

        remove_child(child_id) {
            this.model.children_id = this.model.children_id.filter(id => id !== child_id);
            for (const content of this.get_children_content()) {
                if (content.get_controller().model.get_id() === child_id) {
                    if (child_id.startsWith('ContentGroup')) {
                        content.get_controller().delete_recursive_from_key_value_storage();
                    }
                    content.get_controller().delete();
                    break;
                }
            }
            this.update({ children_id: this.model.children_id });
            return this;
        }

        get_children_content_recursive() {
            const results = [];
            for (const [content] of this.yield_children_content_recursive()) {
                results.push(content);  // Array of Model4LLM.AbstractContent
            }
            return results;
        }
    };

    static TextContentController = class extends Controller4LLM.AbstractContentController {
        constructor(store, model) {
            super(store, model);
            this.model = model;  // Model4LLM.TextContent
            this._store = store;  // LLMstore
        }
    };

    static EmbeddingContentController = class extends Controller4LLM.AbstractContentController {
        constructor(store, model) {
            super(store, model);
            this.model = model;  // Model4LLM.EmbeddingContent
            this._store = store;  // LLMstore
        }

        get_data_raw() {
            return super.get_data_raw().slice(1, -1).split(',').map(Number);
        }

        get_data_rLOD0() {
            return this.get_data_raw().filter((_, index) => index % 10 === 0);
        }

        get_data_rLOD1() {
            return this.get_data_raw().filter((_, index) => index % 100 === 0);
        }

        get_data_rLOD2() {
            return this.get_data_raw().filter((_, index) => index % 1000 === 0);
        }

        get_target() {
            return this._store.find(this.model.target_id);  // Model4LLM.AbstractContent
        }

        update_data_raw(embedding) {
            super.update_data_raw(JSON.stringify(embedding));
            return this;
        }
    };

    static FileLinkContentController = class extends Controller4LLM.AbstractContentController {
        constructor(store, model) {
            super(store, model);
            this.model = model;  // Model4LLM.FileLinkContent
            this._store = store;  // LLMstore
        }
    };

    static BinaryFileContentController = class extends Controller4LLM.AbstractContentController {
        constructor(store, model) {
            super(store, model);
            this.model = model;  // Model4LLM.BinaryFileContent
            this._store = store;  // LLMstore
        }

        read_bytes(filepath) {
            const fs = require('fs');
            return fs.readFileSync(filepath);
        }

        b64decode(file_base64) {
            return Buffer.from(file_base64, 'base64');
        }

        get_data_rLOD0() {
            throw new Error('Binary file has no LOD concept');
        }

        get_data_rLOD1() {
            throw new Error('Binary file has no LOD concept');
        }

        get_data_rLOD2() {
            throw new Error('Binary file has no LOD concept');
        }
    };

    static ImageContentController = class extends Controller4LLM.BinaryFileContentController {
        constructor(store, model) {
            super(store, model);
            this.model = model;  // Model4LLM.ImageContent
            this._store = store;  // LLMstore
        }

        decode_image(encoded_string) {
            const sharp = require('sharp');
            return sharp(Buffer.from(encoded_string, 'base64'));
        }

        get_image() {
            const encoded_image = this.get_data_raw();
            return encoded_image ? this.decode_image(encoded_image) : null;
        }

        get_image_format() {
            const image = this.get_image();
            return image ? image.metadata().format : null;
        }

        get_data_rLOD(lod = 0) {
            const image = this.get_image();
            const ratio = 10 ** (lod + 1);
            if (image.width / ratio <= 0 || image.height / ratio <= 0) {
                throw new Error(`Image size (${image.width}, ${image.height}) of LOD${lod} is too small`);
            }
            return image.resize(Math.floor(image.width / ratio), Math.floor(image.height / ratio));
        }

        get_data_rLOD0() {
            return this.get_data_rLOD(0);
        }

        get_data_rLOD1() {
            return this.get_data_rLOD(1);
        }

        get_data_rLOD2() {
            return this.get_data_rLOD(2);
        }
    };
}
